### PR TITLE
#162725261 Implement commenting on articles

### PR DESCRIPTION
**Pull request by MuhweziDeo**
#### What does this PR do?
Enable commenting on articles

#### Description of Task to be completed?
- create a comment model class to handle CRUD operations on Comments
- create a serailzer class for the model Comments
- create views to handle all CRUD operations for Comments Model
- add routes for the comment views
 
#### How should this be manually tested?
- Fetch and check out this branch with `git fetch origin ft-comment-on-articles-162725261 and git checkout ft-comment-on-articles-162725261` 
- Create and activate your virtual environment with `python3 -m virtualenv <preferredname>` and `source <preferredname>/bin/activate`
- run migrations with `python manage.py migrate comments`
- Start the app by running `python manage.py runserver`
- To comment on an article perform a post on `localhost:8000/api/articles/<slug>/comments` passing json in data `{"body":"body data"}` **requires authentication**
- To view all comments perform a get method on `localhost:8000/api/articles/<slug>/comments`
- To view a single article perform a get request on `localhost:8000/api/articles/<slug>/comments/pk/`
- To update article comment peform a put on `localhost:8000/api/articles/<slug>/comments/pk/` to **be done by person who created the comment**
- To delete article comment perform a delete on `localhost:8000/api/articles/<slug>/comment/pk/` to **be done by person who created the comment**
- To reply to a comment perform a `POST` request on end point `localhost:8000/api/articles/<slug>/comment/pk/reply/`
- To view all replies on a comment perform a `GET` on `localhost:8000/api/articles/<slug>/comment/pk/reply/`
- To view,update,delete a specific reply of a user pass in reply_id in `localhost:8000/api/articles/<slug>/comment/pk/reply/<reply_id>/` **but to be able to delete or update a reply you must be the author of the reply**
- To test this branch run locally run `python manage.py test authors/apps/comments/`

#### What are the relevant pivotal tracker stories?
[ #162725261](https://www.pivotaltracker.com/story/show/162725261)

[Run in Postman](https://app.getpostman.com/run-collection/fd09e0fef2dd4b779a00)

**Screenshots**
screen shot 2019-01-22 at 5 11 39 pm
screen shot 2019-01-22 at 5 12 20 pm
screen shot 2019-01-22 at 5 12 41 pm
screen shot 2019-01-22 at 5 12 57 pm
screen shot 2019-01-22 at 5 13 10 pm
screen shot 2019-01-24 at 9 45 59 am
screen shot 2019-01-24 at 9 46 44 am
screen shot 2019-01-24 at 10 06 34 am
screen shot 2019-01-24 at 10 09 38 am
screen shot 2019-01-24 at 10 14 30 am
screen shot 2019-01-24 at 10 14 41 am

### CODE REVIEW

**Comment by mariamiah** [comment]
screen shot 2019-01-23 at 11 59 50 am
@MuhweziDeo cannot post with this url, please change it to /api/articles/slug/comments/

File: authors/apps/articles/urls.py [hunk 1]
```diff
@@ -1,7 +1,9 @@
 from django.urls import path
 from . import views
-
+from authors.apps.comments.views import CommentView,CommentDetailView
 urlpatterns = [
```
Comment (ja-odur) [inline authors/apps/articles/urls.py:5]: @MuhweziDeo Please add a space here

File: authors/apps/articles/urls.py [hunk 2]
```diff
@@ -1,7 +1,9 @@
...
     path('', views.ArticleListCreateView.as_view(), name='article-create'),
-    path('<slug>/', views.ArticleUpdateDeleteView.as_view(), name='article-update')
+    path('<slug>/', views.ArticleUpdateDeleteView.as_view(), name='article-update'),
+    path('<slug>/comments/',CommentView.as_view(),name='comments'),
```
Comment (ja-odur) [inline authors/apps/articles/urls.py]: @MuhweziDeo Move the urls to the comments app. remember the app has to stand alone.

File: authors/apps/authentication/tests_/test_base_class.py [hunk 3]
```diff
@@ -3,8 +3,8 @@
 from rest_framework.test import APIClient
 from django.urls import reverse
 import json
-
-
+from authors.apps.profiles.models import Profile
+from authors.apps.articles.models import Article
```
Comment (ja-odur) [inline authors/apps/authentication/tests_/test_base_class.py]: @MuhweziDeo duplicate code

File: authors/apps/authentication/tests_/test_base_class.py [hunk 4]
```diff
@@ -49,18 +49,51 @@ def setUp(self):
 
             }
         }
+        
+        self.not_author_login_credentials={
```
Comment (ja-odur) [inline authors/apps/authentication/tests_/test_base_class.py]: @MuhweziDeo Please move this to the test data folder

File: authors/apps/authentication/tests_/test_base_class.py [hunk 5]
```diff
@@ -49,18 +49,51 @@ def setUp(self):
...
             "title": "hello worlfd",
             "description": "desctriptuo",
             "body": 'boddydydabagd'
         }
+        self.profile=Profile.objects.get(user=self.verified_user)
+        self.created_article=Article.objects.create(
+            body=' hello world',description='a description',
+            title='a title',author=self.profile
+        )
+        self.comment_data={
```
Comment (ja-odur) [inline authors/apps/authentication/tests_/test_base_class.py]: @MuhweziDeo these too, move them to the test data

File: authors/apps/comments/tests/test_comment_endpoints.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
+from authors.apps.authentication.tests_ import test_base_class
+from django.shortcuts import reverse
+import json
+
+
+class TestCommentEndpoints(test_base_class.BaseTestClass):
+    def test_create_comment_on_article_successed(self):
```
Comment (ja-odur) [inline authors/apps/comments/tests/test_comment_endpoints.py]: @MuhweziDeo the postfix should be `_succeeds`

File: authors/apps/comments/views.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+from authors.apps.utils.custom_permissions.permissions import check_if_is_author
+
+class CommentView(GenericAPIView):
+    """
+    Allows authenticated users to post a comment on an articles 
+    and also view all comments on an article
+    """
+    permission_classes=(IsAuthenticated,)
+    serializer_class=CommentSerializer
+    def post(self,request,slug):
```
Comment (ja-odur) [inline authors/apps/comments/views.py:43]: @MuhweziDeo add space

File: authors/apps/comments/views.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+        """
+        Allows authenticated users can add comments on articles
```
Comment (ja-odur) [inline authors/apps/comments/views.py:45]: @MuhweziDeo define parameters in the doc string

File: authors/apps/comments/views.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+        self.article=get_object_or_404(Article,slug=slug)
+        serializer=self.serializer_class(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        serializer.save(author=self.author,article=self.article)
+        return Response({
+            'comment':serializer.data
+        },status=status.HTTP_201_CREATED)
+    
+    def get(self,request,slug):
+        """
```
Comment (ja-odur) [inline authors/apps/comments/views.py:63]: @MuhweziDeo Here too, define parameters in the doc string

File: authors/urls.py [hunk 10]
```diff
@@ -41,7 +41,7 @@
     path('api/documentation/', schema_view.with_ui('swagger',
                                                    cache_timeout=0), name='api_documentation'),
     path('api/articles/', include(('authors.apps.articles.urls','articles'), namespace='articles')),
-    path('api/profiles/', include(('authors.apps.profiles.urls','profiles'),  namespace='profiles'))
+    path('api/profiles/', include(('authors.apps.profiles.urls','profiles'),  namespace='profiles')),
```
Comment (ja-odur) [inline authors/urls.py:44]: @MuhweziDeo Please dont edit files not related to the task

**Comment by ja-odur** [comment]
@MuhweziDeo Update your documentation especially the `urls` to working ones.

**Comment by ja-odur** [comment]
@MuhweziDeo Please remove installing requirements from the test section since you have not added any dependency to the requirement.txt file.

**Review by ja-odur — APPROVED** [review]
LGTM

**Review by d-rita — APPROVED** [review]
LGTM

**Review by calvinpete — APPROVED** [review]
Tested and works fine.

**Pull request merged**